Scripting Engine Implementation

bergabman opened the pull request:
I started to implement the scripting engine how I thought it could work smooth. We can make this structure dynamic enugh to fill in all the necessary info we talked about before, like `core_approved` and description of the script. My idea would be to make the interpreted language files in a self contained way, which means the first 4-5 lines of the file could contain the tags in comment space for the interpreter, and the rest of the file can be the script itself.
I'm not sure how we should handle binary files, if the user want's to run a binary we have to store the tags on a different location.

For a main rustscan_scripting config file we could use toml format, where we store some basic variables in case we have any. To read the scripting config, we could add a new arg like `--scripting`

```
tags = ["core_approved", ""]


# optional field, if it's given then only those scripts will run which has a tag ports = [80]
# 
# ex.: 
# ports = [80] 
# ports = [80,81,8080]
ports = [80]


# optional field, if it's given, only that(those) interpreter(s) will be used 
# 
interpreters = ["python3", "bash"]
```

Please check it out how it looks and give comments how to imporve, or if it's even a good idea to do it this way. @bee-san @bernardoamc @CMNatic

bergabman commented:
Right now it works exactly the same, nmap runs as before just with the new structure. So this is actually a working version, just for the record. :)

bergabman commented:
#81 the issue

bee-san commented on src/main.rs:
```diff
@@ -14,6 +14,9 @@ use port_strategy::PortStrategy;
 use benchmark::{Benchmark, NamedTimer};
 
+mod scripts;
+use scripts::{Executable::Nmap, Script};
```
Would we have to add every core script here?

bergabman replied:
No this is just for maybe the ones we want to have built in. Like a default nmap scan. The rest of the scripts will be dynamic.. Maybe we don't need anything like that at all.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+
+// struct ScriptEngine {
+//     nmap: Bin,
+//     python: Bin,
```
We should allow them, in the config file, to specify a file path for Python too :) 

We're going to get a lot of comments like "Python doesn't work" when their default `python` is 2.7 and not `3.7`. Also, we'd need to make sure scripts are compatible with lower version of Python, or to have a version of python mentioned in the script name. 

So if the script tag is `python3.7` it doesn't run if the user has `3.6.9`

bergabman replied:
Good idea, we can probably use absolute path, but if not we can change the working directory of RustScan on the fly and execute from the right directory with the right version of the interpreter.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+//     }
+// }
+
+pub struct Script {
```
This is good, I can't imagine anything else we'd want to add.

Perhaps a specific argument for the language that will execude it?

```rust
language: String
```

So we can easily see if it's Python, Perl, Shell etc

bergabman replied:
Parsed from the file itself, so it stay self-contained. We will have more possibilities to identify the interpreter, so it will be easy to fill that field.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+            ip: ip,
+            ports: ports,
+
+            /*  script tags stored in file header? We can use toml to read it into a struct after parsing the headers manually first
```
We should actually do something like this, ideally:

```
#! Shebang thingy here
# RSE START
# Tags: Core, HTTP, Web, HTTPS
# Ports: 80, 443
# RSE END
```

That way we don't have to parse "maybe line 1 is a shebang?" and "does their comments end after 5 lines or 4?"

Also, if we use TOML like structure within them we can easily parse for the struct.

And searching for `RSE START` and `RSE END` are easier.

Use "end" instead of "stop" because "stop" is physically closer to "start" (same first 2 letters, small minor thing but it might help people read the code)

bergabman replied:
We could do that, or just use a struct with predefined `Option`al fields, and parse from toml to that struct.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+            example:
+
+            #!/bin/python or #!/bin/perl or #!/bin/sh or #!/bin/python3
+            # tags = [core_approved, http, robotstxt, hyper, get_request]
```
Ah you have  this structure here too, nice!

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+        let process = Exec::cmd(self.command).args(&self.arguments);
+        match process.capture() {
+            Ok(c) => {
+                let es = match c.exit_status {
```
Python scripts, by default, do not exit with OS status codes by the way -- we'd have to enforce this with the users submissions :)

bergabman replied:
I think as long as it completes without error, it will return a `0`, every program ends with an exit systemcall, even if it is the interpreter. We should make some tests to see how it reacts to success and errors and what gets returned on the end. I will do that.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,162 @@
+    }
+}
+
+struct ScriptConfig {
```
I was thinking of moving the config and scripts to `~/.rustscan/` so:

```
~/.rustscan/config.toml
~/.rustscan/scripts/nmap.py
```

bergabman replied:
The ScriptConfig function now accepts a base path.

bee-san reviewed (commented):
Ideally, scripts should also be able to access the config file. I think that'd open up a significant amount of customisation.

Here's my idea:

```
/// Finds the directory of where the scripts are stored
/// and returns a vector of fully qualified paths to each script
fn find_scripts{

}

/// Given a vector of scripts, store the tags of each script
/// in a pre-defined struct so we can easily search / sort them later
/// Some tags include:
/// * Language
/// * Developer
/// * Core Approved
/// * How to call the script, so `python3 {locatinon} --ip {ip} --port {port}
/// Our script will fill in the {} for them
fn get_tags{

}

/// Given structs and the tag system
/// Calculate what the tag does
/// I.E the tag(s) might be "(Popular and Core) not Port 80"
/// This returns a vector of paths (or the structs) that meet the tag requirement
/// We will also need to check compatiability with script.
/// If script is Python, do we have Python on the system?
/// If script requires port 80, do we have port 80 open?
/// If the script needs a host name instead of IP, do we have that?
/// You get the idea -- we calculate them instead of the script because
/// multiple scripts may want Python, so we will only need to calculate it once
/// also reduces complexity for script creators
fn calculate_tags_get_scripts{

}

/// Runs the scripts according to how the user says to run them
fn run_scripts{

}
```

bee-san commented:
Can you resolve comments after they're done? and re-request code reviews when you need them too :D

bee-san commented on Cargo.toml:
```diff
@@ -35,6 +35,10 @@ serde = "1.0.116"
+which = "4.0.2"
+anyhow = "1.0.32"
+subprocess = "0.2.6"
+text_placeholder = { version = "0.3", features = ["struct_context"] }
```
Kaeros will love this :)

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+
+    let mut scripts_to_run: Vec<ScriptFile> = Vec::new();
+
+    // Only Scripts that contain all the tags found in ScriptConfig will be selected.
```
By tag do you mean stuff like "core" etc? or do you mean stuff like the author field too? If it's strictly tags like "core, http" etc i can see this being a neat lil security feature (an intended bug)

bergabman replied:
The logic only collect those ScriptFiles from the scripts folder which contains all the tags predefined in the script_config file. For example you can give a crazy unique tag for a script and call that in the ScriptConfig file so only that script will run. We can build more logic around script parsing.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+        if let Ok(file) = File::open(script) {
+            for line in io::BufReader::new(file).lines().skip(1) {
+                if let Ok(mut line) = line {
+                    if line.starts_with("#") && line.len() > 3 {
```
I take it we are only doing Python / Shell scripts, assuming that the lines start with `#`?
It's good for an MVP, but we need to be aware of this bug in the future :)

bergabman replied:
Python, perl, bash all uses `#` as comment, but we for sure can use an extra variable in the ScriptConfig to define that character. That way it will be completely dyunamic, the users could add own characters.

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+#tags = ["core_approved", "bergabman"]
+#developer = [ "testdev", "https://example.org", "https://twitter" ]
+#execute_with = "python3"
+#join_ports = ","
```
what does `","` mean here?

bergabman replied:
If that field is there, the found ports will be joined with that character and passed like that to the script. Like nmap uses `-p 80,8080,9001`

bee-san commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/perl
+#tags = ["core_approved", "bergabman"]
+#developer = [ "testdev", "https://example.org", "https://twitter" ]
+#execute_with = "python3"
```
Why is this `#!/usr/bin/perl` but exex with python3? the script is in perl but the actual thing it's calling is python3?

bergabman replied:
My mistake :D

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+use subprocess::{Exec, ExitStatus};
+use text_placeholder::Template;
+
+pub fn init_scripts() -> Vec<ScriptFile> {
```
There is a common pattern in Rust and that is to initiate a struct resource through a method called `new`. So for this scenario we would obtain the scripts like we are already doing and inside `parse_scripts` we would call `ScriptFile::new()` that would contain the logic to parse this script. Let me know if this makes sense to you. 

```rust

```

bergabman replied:
I implemented a `ScriptFile::new()` function, and moved the parsing logic to there. So the ScriptFile gets parsed and return an `Option`, and gets filtered on `Some()` and `None`. We can expand the information we give to the user in the future to be more clear if sommething goes wrong. Now it just gets discarded.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+}
+
+/*
+Example ScriptFile headers, all fields are optional but we have to do the checks if something is necessary, for example an executor path.
```
Do we need to move this to `ScriptFile` instead of `ScriptConfig` and document `ScriptConfig` itself?

bernardoamc commented on src/input.rs:
```diff
@@ -105,17 +105,21 @@ pub struct Opts {
     #[structopt(long, possible_values = &ScanOrder::variants(), case_insensitive = true, default_value = "serial")]
     pub scan_order: ScanOrder,
 
+    /// Use the top 1000 ports.
+    #[structopt(long)]
+    pub top: bool,
```
Is this part of this PR?

bergabman replied:
This was part of an earlier PR, I just changed up the location of it.

bernardoamc commented on src/main.rs:
```diff
@@ -128,27 +133,45 @@ fn main() {
+            );
+
+            // This part allows us to add commandline arguments to the Script call_format, appending them to the end of the command.
+            if &opts.command.len() > &0 {
```
Can we do `opts.command.len() > 0` here or this consumes the `command`?

bernardoamc commented on src/main.rs:
```diff
@@ -128,27 +133,45 @@ fn main() {
+
+            // This part allows us to add commandline arguments to the Script call_format, appending them to the end of the command.
+            if &opts.command.len() > &0 {
+                let user_nmap_args: Vec<String> = shell_words::split(&opts.command.join(" "))
```
This is not necessarily `nmap` anymore, is it?

bergabman commented on src/main.rs:
```diff
@@ -128,27 +133,45 @@ fn main() {
         }
         detail!("Starting Nmap", opts.greppable, opts.accessible);
```
I changed it to `Starting Script(s)`, that fits better, but we can also remove it of course.

bergabman commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+        return Ok(files_vec);
+    } else {
+        return Err(anyhow!("cant find scripts folder"));
+        // println!("Do you want me to create the {} directory?", &scripts_dir.to_string_lossy());
+        // std::fs::create_dir(&scripts_dir).unwrap();
```
They are gone

bergabman commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+
+pub fn find_scripts(mut path: PathBuf) -> Result<Vec<PathBuf>> {
+    path.push(".rustscan_scripts");
+    // check if dir is available
```
Gone

bergabman commented:
After some discussions on discord, we are getting close to the final idea, (or maybe already there). Please check out the new implementation with examples and let me know what you think. @bee-san @bernardoamc 

It lacks some testing, hopefully we can make a Hacktoberfest issue from them.

bergabman commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    pub fn run(self) -> Result<String> {
```
With the current structure I do consume part of the script so it's necessary as it is now.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,318 @@
+    ip: IpAddr,
+
+    // Ports found with portscan.
+    ports: Vec<u16>,
```
What do you think of calling this `open_ports` since we now know they are open and calling the `port` option something like: `conditional_port` or `port_trigger`? Something that indicates that this port is the condition for the script to run.

bergabman replied:
I changed the ports to open_ports so it's more descriptive, but I'm not sure in the conditional_port or port_trigger. What would you say for predefined_port? 🤔

bergabman replied:
I changed the variable name in the script files and the struct to `trigger_port`, I think that describes it good.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,318 @@
+    tags: Option<Vec<String>>,
+
+    // The format how we want the script to run.
+    call_format: Option<String>,
```
Shouldn't a script always have a `call_format`?

bergabman replied:
I was thinking on some error handling logic after parsing the script, but you are right, it shouldn't be an `Option`, and we can rely on the toml parser to throw an error. Now we just don't run the scripts simply that's not conform in any way. But we can stop it at parsing.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,318 @@
+        ip: IpAddr,
+        ports: Vec<u16>,
+        port: Option<String>,
+        join_ports: Option<String>,
```
Should we call this `ports_separator`?

bergabman replied:
`port_separator` is probably a better name, sounds ok, I'll change it.

bernardoamc commented on src/input.rs:
```diff
@@ -199,6 +201,7 @@ pub struct Config {
     ulimit: Option<rlimit::rlim>,
     scan_order: Option<ScanOrder>,
     command: Option<Vec<String>>,
+    scripts: Option<String>,
```
Since we know what to expect for these scripts we could use an `enum` here with the three options we expect. What do you think?

bernardoamc replied:
This has the advantage that Rust will be able to enforce that we are always taking into account he three possible variants in our code.

bergabman replied:
All right, we can build an enum around the 3 values, now I just handle it like [ &_ => Ok(scripts_to_run),](https://github.com/RustScan/RustScan/pull/242/files#diff-5062b06fea1a548c03f86fb29d0fa7d9R103) :)

bergabman replied:
Also, the commandline parser allows only the 3 options 
`#[structopt(long, possible_values = &["none", "default", "custom"], case_insensitive = true, default_value = "default")]`

bergabman commented:
I made the adjustments we discussed previously, changed `ports` we get from the scanner to `open_ports`, changed `join_ports` in structs and the right places in the example files to `ports_separator`. Created an Enum to parse `--scripta` user options from commandline and configfile with only those 3 values available, the same way `ScanOrder` is implemented. It was an easy copy thanks to the maker of that enum. Expanded the explanation part regarding the `execute_format` section.

I know somebody was talking about the placement of the script_config file, and also renaming it. But I can't seem to find that comment.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,335 @@
+                        if config_hashset.is_subset(&script_hashset) {
+                            scripts_to_run.push(script.to_owned());
+                        }
+                        debug!(
+                            "\nScript tags does not match config tags {:?} {}",
+                            &script_hashset,
+                            script.path.clone().unwrap().display()
+                        );
```
I think this is always running despite if there's a match or not.

bergabman replied:
Yes, I forget to wrap it back to the else statement. Good catch, fixed it.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,335 @@
+        debug!("run self {:?}", &self);
+
+        let mut ports_str = String::new();
+        if let Some(ports_separator) = self.ports_separator {
+            let ports_str_vec: Vec<String> =
+                self.open_ports .iter().map(|port| port.to_string()).collect();
+            ports_str = ports_str_vec.join(&ports_separator);
+        } else if let None = self.ports_separator {
+            let ports_str_vec: Vec<String> =
+                self.open_ports .iter().map(|port| port.to_string()).collect();
+            ports_str = ports_str_vec.join(",");
```
Here we can avoid the `if/else if` by doing:

```rust
let separator = self.ports_separator.unwrap_or(",");
let ports_str = self.open_ports .iter()
  .map(|port| port.to_string())
  .collect()
  .join(&ports_separator);
```

bergabman replied:
It needed some type at `collect::<Vec<String>>()`, but it looks much better. Thanks!

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,335 @@
+            let ports_str_vec: Vec<String> =
+                self.open_ports .iter().map(|port| port.to_string()).collect();
+            ports_str = ports_str_vec.join(",");
+        } else if let Some(port) = self.port {
+            ports_str = port;
+        } else {
+            return Err(anyhow!("Failed to parse ports_separator in scriptfile."));
+        }
```
Do we ever reach those since we are already taking into account the `Some` and `None` cases of `self.ports_separator`?

bergabman replied:
The separator got simplified to `let separator = self.ports_separator.unwrap_or(",".into());` based on your other comment. So now we can get here.

bernardoamc commented on src/scripts/mod.rs:
```diff
@@ -0,0 +1,335 @@
+
+        debug!("\nTo run {}", to_run);
+
+        let arguments: Vec<String> = to_run.split(" ").map(|arg| arg.to_string()).collect();
```
We should probably use `shell_words::split` here instead of a normal split by whitespace to avoid edge cases of argument parsing.

bergabman replied:
I changed to use the original `shell_words::split` parser for the commandline args.

bernardoamc commented on Cargo.toml:
```diff
@@ -36,6 +36,9 @@ serde_derive = "1.0.116"
 trust-dns-resolver = { version = "0.19.5", features = ["dns-over-rustls"] }
+anyhow = "1.0.32"
+subprocess = "0.2.6"
+text_placeholder = { version = "0.3", features = ["struct_context"] }
```
Now we have a version `0.4` that uses less memory 😄

bergabman replied:
Nice! I bumped it. :)

bernardoamc commented on fixtures/test_rustscan_scripts.toml:
```diff
@@ -0,0 +1,14 @@
+# Test/Example ScriptConfig file
+
+# Tags to filter on scripts. Only scripts containing all these tags will run.
+tags = ["core_approved", "example"]
```
Are we over-complicating this tags system? Maybe we can ignore `tags` and `developer` completely until we actually allow people to contribute scripts. This will keep our code simpler and allow us to evolve it when we have the need to.

bernardoamc commented on fixtures/test_script.py:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/python3
+#tags = ["core_approved", "example",]
+#developer = [ "example", "https://example.org" ]
+#port = "80"
+#call_format = "python3 {{script}} {{ip}} {{port}}"
```
Now that I'm thinking about it we should probably make those independent of the script itself since we are preventing other scripts that do not support `#` as a comment from running.

bergabman replied:
For the first release, I think we are good with accepting the files which do accept that as a comment. We can of course expand the parsing of ScriptFiles to include other marks.

bee-san commented:
@bergabman conflicts! :( But other than conflicts, how's this coming? Are we ready to release? Do you need any help on anything? :heart:

bernardoamc commented:
There are a few details/performance improvements that we can tackle, but if we are happy with the functionality we can always merge this and keep iterating on it. Great job on this @bergabman <3 

The things that we can tackle later on:

1. Reduce memory allocation through usage of references and lifetimes
    * Strings
    * Hashset
2. Keep improving the scripts usage through things like port matching and so on

bergabman commented on src/main.rs:
```diff
@@ -112,46 +118,71 @@ fn main() {
+                Ok(script_result) => {
+                    detail!(
+                        format!("{}", script_result),
+                        opts.greppable.clone(),
+                        opts.accessible.clone()
```
No we don't, for some reason I ended up cloning them during development. I think the struct itself didn't have `Copy` implemented at the beginning. Fixed both.

bergabman commented on src/main.rs:
```diff
@@ -165,7 +196,7 @@ fn print_opening(opts: &Opts) {
 | .-. \| {_} |.-._} } | |  .-._} }\     }/  /\  \| |\  |
 `-' `-'`-----'`----'  `-'  `----'  `---' `-'  `-'`-' `-'
-Faster Nmap scanning with Rust."#;
+The Modern Day Port Scanner."#;
```
I know right :)

bergabman commented on src/main.rs:
```diff
@@ -112,46 +118,71 @@ fn main() {
+                Err(e) => {
+                    warning!(
+                        &format!("Error {}", e.to_string()),
+                        opts.greppable.clone(),
+                        opts.accessible.clone()
```
You are right, see other comment.

bergabman commented on src/main.rs:
```diff
@@ -112,46 +118,71 @@ fn main() {
-            .spawn()
-            .expect("failed to execute nmap process");
+            // This part allows us to add commandline arguments to the Script call_format, appending them to the end of the command.
+            if opts.command.len() > 0 {
```
I think it makes sense, in case the user want to simply append an extra argument to the execution of the scripts. For example in case of an nmap script, we give the possibility to just append extra arguments to RustScan that gets passed to the script. An extra `-sC` switch for nmap will be just a quick append to the RustScan commandline arguments. If we remove it, the users will have to change the script file every time they want to append a new argument to it. This will also improve development time and experience for writing RustScan script files, as the writer of the script can test arguments in terminal and append them to the script execution format if they fit.

bergabman commented:
I improved the PR based on the reviews. I think this version of the ScriptingEngine will be a good start to let people write custom scripts for RustScan. It certainly can be improved and as @bernardoamc said we should tackel efficiency points like using references and lifetimes. Luckily we don't have to parse a lot of script files as of now, and I can imagine people will most likely run 1 script with the results of the portscan, so memory allocations are not that big of a deal. But to have a more elegant and correct code we will have to tackle those points. 

The module lacks testing, but the example files in the `fixtures` directory can be used to write the tests. There is an example script config file and 4 different scripts to work with. Maybe it's worth to create an issue to ask people to write tests for us :).

bee-san commented:
Considering this issue is for an initial RSE, I see no reason why this shouldn't be merged :) The most important thing now should be writing docs on how to use the thing :sweat_smile:
